crates/table/src/table.rs

```diff
@@ -234,20 +250,242 @@ impl Table {
+        // SAFETY: Per post-condition of `confirm_insertion`, `row_ptr` refers to a valid row.
+        let row_ref = unsafe { self.inner.get_row_ref_unchecked(blob_store, row_ptr) };
+        Ok((hash, row_ref))
+    }
+
+    /// Insert a `row` into this table during replay.
+    ///
+    /// NOTE: This method skips index updating. Use `insert` to insert a row with index updating.
+    pub fn insert_for_replay(
+        &mut self,
+        blob_store: &mut dyn BlobStore,
+        row: &ProductValue,
+    ) -> Result<(RowHash, RowPointer), InsertError> {
+        // Insert the `row`. There should be no errors
+        let (row_ref, blob_bytes) = self.insert_physically_pv(blob_store, row)?;
+        let row_ptr = row_ref.pointer();
 
-        // SAFETY: We just inserted `ptr`,
-        // and `insert_into_indices` didn't remove it,
-        // so it must be present.
+        // SAFETY: We just inserted the row, so `self.is_row_present(row_ptr)` holds.
+        let row_hash = unsafe { self.insert_into_pointer_map(blob_store, row_ptr) }?;
+
+        self.update_statistics_added_row(blob_bytes);
+
+        Ok((row_hash, row_ptr))
+    }
+
+    /// Physically inserts `row` into the page
+    /// without inserting it logically into the pointer map.
+    ///
+    /// This is useful when we need to insert a row temporarily to get back a `RowPointer`.
+    /// A call to this method should be followed by a call to [`delete_internal_skip_pointer_map`].
+    pub fn insert_physically_pv<'a>(
+        &'a mut self,
+        blob_store: &'a mut dyn BlobStore,
+        row: &ProductValue,
+    ) -> Result<(RowRef<'a>, BlobNumBytes), InsertError> {
+        // SAFETY: `self.pages` is known to be specialized for `self.row_layout`,
+        // as `self.pages` was constructed from `self.row_layout` in `Table::new`.
+        let (ptr, blob_bytes) = unsafe {
+            write_row_to_pages(
+                &mut self.inner.pages,
+                &self.inner.visitor_prog,
+                blob_store,
+                &self.inner.row_layout,
+                row,
+                self.squashed_offset,
+            )
+        }?;
+        // SAFETY: We just inserted `ptr`, so it must be present.
         let row_ref = unsafe { self.inner.get_row_ref_unchecked(blob_store, ptr) };
-        Ok((hash, row_ref))
+
+        Ok((row_ref, blob_bytes))
+    }
+
+    /// Physically insert a `row`, encoded in BSATN, into this table,
+    /// storing its large var-len members in the `blob_store`.
+    ///
+    /// On success, returns the hash of the newly-inserted row,
+    /// and a `RowRef` referring to the row.
+    ///
+    /// This does not check for set semantic or unique constraints.
+    ///
+    /// This is also useful when we need to insert a row temporarily to get back a `RowPointer`.
+    /// In this case, A call to this method should be followed by a call to [`delete_internal_skip_pointer_map`].
```

Comment:
Can you expand this comment with a short paragraph about what will happen if this is called with a `row` that is not valid BSATN at the table's row type? Doesn't have to be super in-depth, but just to set our expectations between "Precisely detects all type errors" and "corrupts the table and burns down your house."